helper/resource/testing_new.go

```diff
@@ -0,0 +1,615 @@
+package resource
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"reflect"
+	"strconv"
+	"strings"
+	"testing"
+
+	tftest "github.com/apparentlymart/terraform-plugin-test"
+	"github.com/davecgh/go-spew/spew"
+	tfjson "github.com/hashicorp/terraform-json"
+	"github.com/hashicorp/terraform-plugin-sdk/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/internal/addrs"
+	"github.com/hashicorp/terraform-plugin-sdk/internal/tfdiags"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+)
+
+func shimAttributeValues(flatmap map[string]string, currentKey string, value interface{}) {
```

Comment:
In general a second set of eyes on the shimming/recursion would be good, likely from the Core team? I just winged this (unless you thoroughly also have investigated this already and made fixes and/or found it good to go)